In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
import re

# Data Tidying and Cleaning Lab
## Reading, tidying and cleaning data. Preparing data for exploration, mining, analysis and learning

In this lab, you'll be working with the Coffee Quality Index dataset, located [here](https://www.kaggle.com/datasets/volpatto/coffee-quality-database-from-cqi). For convenience (and to save trouble in case you can't download files, or someone uploads a newer version), I've provided the dataset in the `data/` folder. The metadata (description) is at the Kaggle link. For this lab, you'll only need `merged_data_cleaned.csv`, as it is the concatenation of the other two datasets.

In this (and the following labs), you'll get several questions and problems. Do your analysis, describe it, use any tools and plots you wish, and answer. You can create any amount of cells you'd like.

Sometimes, the answers will not be unique, and they will depend on how you decide to approach and solve the problem. This is usual - we're doing science after all!

It's a good idea to save your clean dataset after all the work you've done to it.

### Problem 1. Read the dataset (1 point)
This should be self-explanatory. The first column is the index.

In [5]:
coffe_index = pd.read_csv("data/merged_data_cleaned.csv", index_col=0)

In [6]:
coffe_index

Species                     Owner Country.of.Origin  \
0     Arabica                 metad plc          Ethiopia   
1     Arabica                 metad plc          Ethiopia   
2     Arabica  grounds for health admin         Guatemala   
3     Arabica       yidnekachew dabessa          Ethiopia   
4     Arabica                 metad plc          Ethiopia   
...       ...                       ...               ...   
1334  Robusta               luis robles           Ecuador   
1335  Robusta               luis robles           Ecuador   
1336  Robusta               james moore     United States   
1337  Robusta             cafe politico             India   
1338  Robusta             cafe politico           Vietnam   

                                     Farm.Name Lot.Number            Mill  \
0                                    metad plc        NaN       metad plc   
1                                    metad plc        NaN       metad plc   
2     san marcos barrancas "san cristobal cuch        NaN             NaN   
3        yidnekachew dabessa coffee plantation        NaN         wolensu   
4                                    metad plc        NaN       metad plc   
...                                        ...        ...             ...   
1334                                 robustasa   Lavado 1     our own lab   
1335                                 robustasa   Lavado 3  own laboratory   
1336                           fazenda cazengo        NaN    cafe cazengo   
1337                                       NaN        NaN             NaN   
1338                                       NaN        NaN             NaN   

             ICO.Number                                Company       Altitude  \
0             2014/2015      metad agricultural developmet plc      1950-2200   
1             2014/2015      metad agricultural developmet plc      1950-2200   
2                   NaN                                    NaN  1600 - 1800 m   
3                   NaN  yidnekachew debessa coffee plantation      1800-2200   
4             2014/2015      metad agricultural developmet plc      1950-2200   
...                 ...                                    ...            ...   
1334                NaN                              robustasa            NaN   
1335                NaN                              robustasa             40   
1336                NaN                global opportunity fund     795 meters   
1337  14-1118-2014-0087                          cafe politico            NaN   
1338                NaN                          cafe politico            NaN   

                             Region  ...       Color  Category.Two.Defects  \
0                      guji-hambela  ...       Green                     0   
1                      guji-hambela  ...       Green                     1   
2                               NaN  ...         NaN                     0   
3                            oromia  ...       Green                     2   
4                      guji-hambela  ...       Green                     2   
...                             ...  ...         ...                   ...   
1334               san juan, playas  ...  Blue-Green                     1   
1335               san juan, playas  ...  Blue-Green                     0   
1336  kwanza norte province, angola  ...         NaN                     6   
1337                            NaN  ...       Green                     1   
1338                            NaN  ...         NaN                     9   

               Expiration                  Certification.Body  \
0         April 3rd, 2016  METAD Agricultural Development plc   
1         April 3rd, 2016  METAD Agricultural Development plc   
2          May 31st, 2011        Specialty Coffee Association   
3        March 25th, 2016  METAD Agricultural Development plc   
4         April 3rd, 2016  METAD Agricultural Development plc   
...                   ...                       

### Problem 2. Observations and features (1 point)
How many observations are there? How many features? Which features are numerical, and which are categorical?

**Note:** Think about the _meaning_, not the data types. The dataset hasn't been thoroughly cleaned.

#### Observations 
In a dataset observation is a single row of data and represent one individual point of record. In out dataset observations are equal to the number rows - 1340. 
#### Features
The features are specific measurable properties or characteristics of the data and they are equal to the columns. In the our dataset we have 43 features.
##### Numerical Features
These are features that represent quantities.Usually,they are numbers with which we can do mathematical operations.For example in our dataset those are ICO.Number, Alttitude, Category.Two.Defects, etc.
##### Categorical Features
These are features that represent categories or groups. They can be nomial or ordinal. In our dataset such features are Species, Owner, Country.Of.Origin, etc.  

### Problem 3. Column manipulation (1 point)
Make the column names more Pythonic (which helps with the quality and... aesthetics). Convert column names to `snake_case`, i.e. `species`, `country_of_origin`, `ico_number`, etc. Try to not do it manually.

In [10]:
coffe_index.columns

Index(['Species', 'Owner', 'Country.of.Origin', 'Farm.Name', 'Lot.Number',
       'Mill', 'ICO.Number', 'Company', 'Altitude', 'Region', 'Producer',
       'Number.of.Bags', 'Bag.Weight', 'In.Country.Partner', 'Harvest.Year',
       'Grading.Date', 'Owner.1', 'Variety', 'Processing.Method', 'Aroma',
       'Flavor', 'Aftertaste', 'Acidity', 'Body', 'Balance', 'Uniformity',
       'Clean.Cup', 'Sweetness', 'Cupper.Points', 'Total.Cup.Points',
       'Moisture', 'Category.One.Defects', 'Quakers', 'Color',
       'Category.Two.Defects', 'Expiration', 'Certification.Body',
       'Certification.Address', 'Certification.Contact', 'unit_of_measurement',
       'altitude_low_meters', 'altitude_high_meters', 'altitude_mean_meters'],
      dtype='object')

In [11]:
# Use a function to convert the column names to lower case and snake_case

coffe_index = coffe_index.rename(columns = lambda col: col.lower().replace(".", "_"))

In [12]:
coffe_index.columns

Index(['species', 'owner', 'country_of_origin', 'farm_name', 'lot_number',
       'mill', 'ico_number', 'company', 'altitude', 'region', 'producer',
       'number_of_bags', 'bag_weight', 'in_country_partner', 'harvest_year',
       'grading_date', 'owner_1', 'variety', 'processing_method', 'aroma',
       'flavor', 'aftertaste', 'acidity', 'body', 'balance', 'uniformity',
       'clean_cup', 'sweetness', 'cupper_points', 'total_cup_points',
       'moisture', 'category_one_defects', 'quakers', 'color',
       'category_two_defects', 'expiration', 'certification_body',
       'certification_address', 'certification_contact', 'unit_of_measurement',
       'altitude_low_meters', 'altitude_high_meters', 'altitude_mean_meters'],
      dtype='object')

### Problem 4. Bag weight (1 point)
What's up with the bag weights? Make all necessary changes to the column values. Don't forget to document your methods and assumptions.

In [14]:
coffe_index.bag_weight

0       60 kg
1       60 kg
2           1
3       60 kg
4       60 kg
        ...  
1334     2 kg
1335     2 kg
1336     1 kg
1337    5 lbs
1338    5 lbs
Name: bag_weight, Length: 1339, dtype: object

In [15]:
coffe_index.bag_weight.dtype

dtype('O')

In [16]:
coffe_index.bag_weight.value_counts()

bag_weight
1 kg        331
60 kg       256
69 kg       200
70 kg       156
2 kg        122
100 lbs      59
30 kg        29
5 lbs        23
6            19
20 kg        14
50 kg        14
10 kg        11
59 kg        10
1 lbs         8
1             7
3 lbs         7
5 kg          7
2 lbs         5
4 lbs         4
80 kg         4
18975 kg      4
0 lbs         3
46 kg         3
29 kg         2
9000 kg       2
25 kg         2
66 kg         2
35 kg         2
12000 kg      2
40 kg         2
6 kg          2
19200 kg      2
15 kg         2
13800 kg      1
100 kg        1
55 lbs        1
4 kg          1
67 kg         1
350 kg        1
3 kg          1
8 kg          1
80 lbs        1
24 kg         1
1500 kg       1
2 kg,lbs      1
0 kg          1
660 kg        1
1218 kg       1
2             1
18 kg         1
150 lbs       1
18000 kg      1
1 kg,lbs      1
132 lbs       1
34 kg         1
130 lbs       1
Name: count, dtype: int64

#### Problems found with bag_weight column
Looking at the value_counts(), the dtypes and the column data we can suggest that there are the following problems 
with the bag_weight column:
- different weigth units (kg and lbs)
- numbers and units are no separated
- entries without units
- ambiguous entries - 1 kg, lbs
- problems with whitespaces and casing
- data type is object

In respect to the different weight units our analysis shows that we have 1198 entries in which the weight is shown in kg, 114 observations in which the weight is shown in lbs, 27 with missing units and 2 with mix entries.The kilogram is used as a measure in 88.15% of the observations.      

To solve above issues with the bag_weight columns we will perform a set of operations starting with the
normalization of all the data in the column

In [19]:
# 1. Normalize casting and spaces 

coffe_index["bag_weight"] = coffe_index["bag_weight"].str.lower().str.replace(",", " ").str.strip()

In [20]:
coffe_index["bag_weight"]

0       60 kg
1       60 kg
2           1
3       60 kg
4       60 kg
        ...  
1334     2 kg
1335     2 kg
1336     1 kg
1337    5 lbs
1338    5 lbs
Name: bag_weight, Length: 1339, dtype: object

In [21]:
# 2. Extract weight
# We want to see what is the clear weight of each observation.
# To do this we will extract the weight as a float.
# The extracted weight we will keep in a new feature column - bag_weight_num.

coffe_index["bag_weight_num"] = coffe_index["bag_weight"].str.extract(r'(\d+\.?\d*)')[0].astype(float)

In [22]:
coffe_index["bag_weight_num"]

0       60.0
1       60.0
2        1.0
3       60.0
4       60.0
        ... 
1334     2.0
1335     2.0
1336     1.0
1337     5.0
1338     5.0
Name: bag_weight_num, Length: 1339, dtype: float64

In [23]:
# 3.Extract the unit

# As a next step we will extract the unit from the bag_weight column and store it in a new column bag_weight_unit
# The mix entries will be converted to kg, because the first entry in both observations is kg. The lbs entries 
# aftert the kg, probabbly is a techincal error. 

coffe_index["bag_weight_unit"] = coffe_index["bag_weight"].str.extract(r'\d+\.?\d*\s*([a-z]*)')[0]

In [24]:
coffe_index["bag_weight_unit"].value_counts()

bag_weight_unit
kg     1198
lbs     114
         27
Name: count, dtype: int64

In [25]:
# Analyze the observations without entries
# We will try to find out from which country they are and check what is the units used in that country
# Replace the missing units with the units used in that perticular country 

coffe_index[~coffe_index['bag_weight_unit'].isin(['kg', 'lbs'])]

species                             owner       country_of_origin  \
2     Arabica          grounds for health admin               Guatemala   
13    Arabica          grounds for health admin  United States (Hawaii)   
17    Arabica       ethiopia commodity exchange                Ethiopia   
26    Arabica       ethiopia commodity exchange                Ethiopia   
30    Arabica                         nora zeas               Nicaragua   
38    Arabica       ethiopia commodity exchange                Ethiopia   
50    Arabica       ethiopia commodity exchange                Ethiopia   
51    Arabica          grounds for health admin  United States (Hawaii)   
82    Arabica       ethiopia commodity exchange                Ethiopia   
94    Arabica       ethiopia commodity exchange                Ethiopia   
118   Arabica       ethiopia commodity exchange                Ethiopia   
119   Arabica                   philip schluter                Ethiopia   
124   Arabica          grounds for health admin               Nicaragua   
143   Arabica       ethiopia commodity exchange                Ethiopia   
190   Arabica            the coffee source inc.               Guatemala   
294   Arabica       ethiopia commodity exchange                Ethiopia   
313   Arabica  kona pacific farmers cooperative  United States (Hawaii)   
395   Arabica          grounds for health admin                  Mexico   
477   Arabica       ethiopia commodity exchange                Ethiopia   
531   Arabica          grounds for health admin               Nicaragua   
606   Arabica  kona pacific farmers cooperative  United States (Hawaii)   
673   Arabica                        adam kline               Guatemala   
933   Arabica  kona pacific farmers cooperative  United States (Hawaii)   
1093  Arabica       ethiopia commodity exchange                Ethiopia   
1129  Arabica       ethiopia commodity exchange                Ethiopia   
1191  Arabica                 masamichi hiroike                  Mexico   
1261  Arabica                 masamichi hiroike                  Mexico   

                                     farm_name lot_number  \
2     san marcos barrancas "san cristobal cuch        NaN   
13                               arianna farms        NaN   
17                                         NaN        NaN   
26                                         NaN        NaN   
30                                         NaN        NaN   
38    phone number | 0911-51-08-01, email | at        NaN   
50             contact | sidacoop@ethionet.net        NaN   
51                      hula daddy kona coffee        NaN   
82                                         NaN        NaN   
94                                         NaN        NaN   
118            contact | sidacoop@ethionet.net        NaN   
119            contact | sidacoop@ethionet.net        NaN   
124                               santa martha        NaN   
143            contact | sidacoop@ethionet.net        NaN   
190                                  las minas        NaN   
294   phone number | 0911-51-08-01, email | at        NaN   
313                                        NaN        NaN   
395                                        NaN        NaN   
477          tega and tula special coffee farm        NaN   
531                                 santa rosa        NaN   
606                                        NaN        NaN   
673                                        NaN        NaN   
933                                        NaN        NaN   
1093                                       NaN        NaN   
1129  phone number | 0911-51-08-01, email | at        NaN   
1191                                   various        NaN   
1261                                   various        NaN   

                                          mill              ico_number  \
2                                          NaN                     NaN   
13                                         NaN  

In [26]:
coffe_index[coffe_index["country_of_origin"] == "United States (Hawaii)"].head(5)

species                             owner       country_of_origin  \
13  Arabica          grounds for health admin  United States (Hawaii)   
36  Arabica  kona pacific farmers cooperative  United States (Hawaii)   
37  Arabica  kona pacific farmers cooperative  United States (Hawaii)   
51  Arabica          grounds for health admin  United States (Hawaii)   
64  Arabica  kona pacific farmers cooperative  United States (Hawaii)   

                           farm_name lot_number mill              ico_number  \
13                     arianna farms        NaN  NaN                     NaN   
36                               NaN        NaN  NaN                  220452   
37  kona pacific farmers cooperative        NaN  NaN  HDOA 210128 Kona Fancy   
51            hula daddy kona coffee        NaN  NaN                     NaN   
64                               NaN        NaN  NaN                  220451   

                             company altitude region  ...          expiration  \
13                               NaN  2000 ft   kona  ...      May 31st, 2011   
36  kona pacific farmers cooperative      NaN   kona  ...     April 6th, 2013   
37                               NaN      NaN   kona  ...  January 28th, 2012   
51                               NaN   1400ft   kona  ...      May 31st, 2011   
64  kona pacific farmers cooperative      NaN   kona  ...     April 6th, 2013   

              certification_body                     certification_address  \
13  Specialty Coffee Association  36d0d00a3724338ba7937c52a378d085f2172daa   
36  Specialty Coffee Association  36d0d00a3724338ba7937c52a378d085f2172daa   
37  Specialty Coffee Association  36d0d00a3724338ba7937c52a378d085f2172daa   
51  Specialty Coffee Association  36d0d00a3724338ba7937c52a378d085f2172daa   
64  Specialty Coffee Association  36d0d00a3724338ba7937c52a378d085f2172daa   

                       certification_contact unit_of_measurement  \
13  0878a7d4b9d35ddbf0fe2ce69a2062cceb45a660                  ft   
36  0878a7d4b9d35ddbf0fe2ce69a2062cceb45a660                  ft   
37  0878a7d4b9d35ddbf0fe2ce69a2062cceb45a660                  ft   
51  0878a7d4b9d35ddbf0fe2ce69a2062cceb45a660                  ft   
64  0878a7d4b9d35ddbf0fe2ce69a2062cceb45a660                  ft   

   altitude_low_meters altitude_high_meters altitude_mean_meters  \
13              609.60               609.60               609.60   
36                 NaN                  NaN                  NaN   
37                 NaN                  NaN                  NaN   
51              426.72               426.72               426.72   
64                 NaN                  NaN                  NaN   

   bag_weight_num  bag_weight_unit  
13            1.0                   
36          100.0              lbs  
37          100.0              lbs  
51            2.0                   
64          100.0              lbs  

[5 rows x 45 columns]

In [27]:
coffe_index[coffe_index["country_of_origin"] == "Mexico"].head(5)

species                                            owner  \
22   Arabica                            roberto licona franco   
99   Arabica                 el equimite, cafetal biodinámico   
103  Arabica                                        young kim   
111  Arabica                            artemio zapata tejeda   
140  Arabica  organizaciones de productores de cafe colimense   

    country_of_origin                           farm_name          lot_number  \
22             Mexico                        la herradura                 NaN   
99             Mexico                             various  Mezcla Veracruz 01   
103            Mexico                           kassandra                 NaN   
111            Mexico                          el suspiro                 NaN   
140            Mexico  el nogal, la higuera, los laureles                 NaN   

                                            mill                   ico_number  \
22                                  la herradura                            0   
99     cafetal el equimite, rancho agroecológico                          NaN   
103                                          NaN                  016-2484-04   
111                                  pacho viejo                   1104385872   
140  cafetaleros de el remudadero s.p.r. de r.l.  703816889, 108282,703816891   

                                               company   altitude  \
22                                                 NaN       1320   
99   cafetal el equimite, rancho agroecológico, s.p...       1300   
103                                                NaN  1400 masl   
111                                                NaN       1200   
140  organizaciones de productores de cafe colimens...       1200   

            region  ...            expiration  \
22          xalapa  ...       July 26th, 2013   
99        veracruz  ...  September 21st, 2017   
103         mexico  ...       April 5th, 2012   
111       coatepec  ...       July 11th, 2013   
140  el remudadero  ...        July 2nd, 2013   

                     certification_body  \
22                              AMECAFE   
99   Centro Agroecológico del Café A.C.   
103        Specialty Coffee Association   
111                             AMECAFE   
140                             AMECAFE   

                        certification_address  \
22   59e396ad6e22a1c22b248f958e1da2bd8af85272   
99   3b8dfdd621590b424ff64e0b76df7d6a92e1c628   
103  36d0d00a3724338ba7937c52a378d085f2172daa   
111  59e396ad6e22a1c22b248f958e1da2bd8af85272   
140  59e396ad6e22a1c22b248f958e1da2bd8af85272   

                        certification_contact unit_of_measurement  \
22   0eb4ee5b3f47b20b049548a2fd1e7d4a2b70d0a7                   m   
99   d470dc009281519e30da6ead1c649fcd7670f386                   m   
103  0878a7d4b9d35ddbf0fe2ce69a2062cceb45a660                   m   
111  0eb4ee5b3f47b20b049548a2fd1e7d4a2b70d0a7                   m   
140  0eb4ee5b3f47b20b049548a2fd1e7d4a2b70d0a7                   m   

    altitude_low_meters altitude_high_meters altitude_mean_meters  \
22               1320.0               1320.0               1320.0   
99               1300.0               1300.0               1300.0   
103              1400.0               1400.0               1400.0   
111              1200.0               1200.0               1200.0   
140              1200.0               1200.0               1200.0   

    bag_weight_num  bag_weight_unit  
22             1.0               kg  
99            30.0               kg  
103            1.0               kg  
111            1.0               kg  
140            1.0               kg  

[5 rows x 45 columns]

In [28]:
coffe_index[coffe_index["country_of_origin"] == "Guatemala"].head(5)

species                      owner country_of_origin  \
2   Arabica   grounds for health admin         Guatemala   
52  Arabica  juan luis alvarado romero         Guatemala   
57  Arabica  juan luis alvarado romero         Guatemala   
58  Arabica  juan luis alvarado romero         Guatemala   
69  Arabica  juan luis alvarado romero         Guatemala   

                                   farm_name  lot_number  \
2   san marcos barrancas "san cristobal cuch         NaN   
52                     la esperanza y anexos  11/23/0252   
57                                valparaiso         NaN   
58                            finca el limon         NaN   
69                                   asoperc         NaN   

                                mill  ico_number               company  \
2                                NaN         NaN                   NaN   
52                  beneficio ixchel  11/23/0252  unex guatemala, s.a.   
57              beneficio valparaiso  11/1886/03              sertrasa   
58                  beneficio serben  11-853-155               unicafe   
69  beneficio exportacafe agua santa   11-52-492            exportcafe   

            altitude         region  ...           expiration  \
2      1600 - 1800 m            NaN  ...       May 31st, 2011   
52               NaN  nuevo oriente  ...       May 16th, 2017   
57              5000     acatenango  ...       July 3rd, 2013   
58              4650  nuevo oriente  ...       June 5th, 2013   
69  5600-5760 metros  huehuetenango  ...  February 26th, 2014   

              certification_body                     certification_address  \
2   Specialty Coffee Association  36d0d00a3724338ba7937c52a378d085f2172daa   
52  Asociacion Nacional Del Café  b1f20fe3a819fd6b2ee0eb8fdc3da256604f1e53   
57  Asociacion Nacional Del Café  b1f20fe3a819fd6b2ee0eb8fdc3da256604f1e53   
58  Asociacion Nacional Del Café  b1f20fe3a819fd6b2ee0eb8fdc3da256604f1e53   
69  Asociacion Nacional Del Café  b1f20fe3a819fd6b2ee0eb8fdc3da256604f1e53   

                       certification_contact unit_of_measurement  \
2   0878a7d4b9d35ddbf0fe2ce69a2062cceb45a660                   m   
52  724f04ad10ed31dbb9d260f0dfd221ba48be8a95                   m   
57  724f04ad10ed31dbb9d260f0dfd221ba48be8a95                  ft   
58  724f04ad10ed31dbb9d260f0dfd221ba48be8a95                  ft   
69  724f04ad10ed31dbb9d260f0dfd221ba48be8a95                  ft   

   altitude_low_meters altitude_high_meters altitude_mean_meters  \
2              1600.00             1800.000             1700.000   
52                 NaN                  NaN                  NaN   
57             1524.00             1524.000             1524.000   
58             1417.32             1417.320             1417.320   
69             1706.88             1755.648             1731.264   

   bag_weight_num  bag_weight_unit  
2             1.0                   
52           69.0               kg  
57            1.0               kg  
58            1.0               kg  
69           69.0               kg  

[5 rows x 45 columns]

In [29]:
coffe_index[coffe_index["country_of_origin"] == "Nicaragua"]

species                                      owner country_of_origin  \
30    Arabica                                  nora zeas         Nicaragua   
124   Arabica                   grounds for health admin         Nicaragua   
217   Arabica                 cooperativa el gorrion r.l         Nicaragua   
317   Arabica  comercial internacional exportadora, s.a.         Nicaragua   
531   Arabica                   grounds for health admin         Nicaragua   
758   Arabica                             bencafe, s. a.         Nicaragua   
822   Arabica                                chris finch         Nicaragua   
875   Arabica  comercial internacional exportadora, s.a.         Nicaragua   
911   Arabica                 exportadora atlantic, s.a.         Nicaragua   
964   Arabica  comercial internacional exportadora, s.a.         Nicaragua   
971   Arabica                 exportadora atlantic, s.a.         Nicaragua   
988   Arabica                             rolando lacayo         Nicaragua   
1023  Arabica                                chris finch         Nicaragua   
1040  Arabica  comercial internacional exportadora, s.a.         Nicaragua   
1061  Arabica                 exportadora atlantic, s.a.         Nicaragua   
1099  Arabica           gloria antonieta escobar urrutia         Nicaragua   
1106  Arabica                 exportadora atlantic, s.a.         Nicaragua   
1116  Arabica                          honor dela fuente         Nicaragua   
1122  Arabica  comercial internacional exportadora, s.a.         Nicaragua   
1141  Arabica           asociación aldea global jinotega         Nicaragua   
1169  Arabica                 cooperativa el gorrion r.l         Nicaragua   
1206  Arabica                 exportadora atlantic, s.a.         Nicaragua   
1214  Arabica                 exportadora atlantic, s.a.         Nicaragua   
1215  Arabica                 exportadora atlantic, s.a.         Nicaragua   
1246  Arabica                             bencafe, s. a.         Nicaragua   
1308  Arabica                 exportadora atlantic, s.a.         Nicaragua   

                             farm_name                   lot_number  \
30                                 NaN                          NaN   
124                       santa martha                          NaN   
217                          las nubes                          NaN   
317                      santa matilde            606118 / K1730101   
531                         santa rosa                          NaN   
758          el brigadum - la coronela                017/105/16042   
822                                NaN                          NaN   
875                      santa matilde            345725 / K1620088   
911                   finca las marías  017-053-0211 / 017-053-0212   
964            cafetales santa matilde                          NaN   
971          las promesas de san blass                 017-053-0134   
988            cafetales santa matilde                          NaN   
1023                               NaN                          NaN   
1040           cafetales santa matilde                          NaN   
1061                    los saltarines                 017-053-0046   
1099                 finca el consuelo                          NaN   
1106                       los mesones                          NaN   
1116                               NaN                          NaN   
1122                     santa matilde            605135 / K1730103   
1141  asociación aldea global jinotega                          NaN   
1169                         las nubes                          NaN   
1206                  finca las marías        017-053-0211/053-0212   
1214                    los saltarines                 017-053-0046   
1215                               NaN                          NaN   
1246         el brigadum - la coronela                017/105/16039   
1308                  finca las marías   017-053-0211

In [30]:
coffe_index[coffe_index["country_of_origin"] == "Ethiopia"].head(5)

species                        owner country_of_origin  \
0  Arabica                    metad plc          Ethiopia   
1  Arabica                    metad plc          Ethiopia   
3  Arabica          yidnekachew dabessa          Ethiopia   
4  Arabica                    metad plc          Ethiopia   
7  Arabica  ethiopia commodity exchange          Ethiopia   

                               farm_name lot_number       mill ico_number  \
0                              metad plc        NaN  metad plc  2014/2015   
1                              metad plc        NaN  metad plc  2014/2015   
3  yidnekachew dabessa coffee plantation        NaN    wolensu        NaN   
4                              metad plc        NaN  metad plc  2014/2015   
7                                  aolme        NaN    c.p.w.e   010/0338   

                                 company   altitude        region  ...  \
0      metad agricultural developmet plc  1950-2200  guji-hambela  ...   
1      metad agricultural developmet plc  1950-2200  guji-hambela  ...   
3  yidnekachew debessa coffee plantation  1800-2200        oromia  ...   
4      metad agricultural developmet plc  1950-2200  guji-hambela  ...   
7                                    NaN  1570-1700        oromia  ...   

            expiration                  certification_body  \
0      April 3rd, 2016  METAD Agricultural Development plc   
1      April 3rd, 2016  METAD Agricultural Development plc   
3     March 25th, 2016  METAD Agricultural Development plc   
4      April 3rd, 2016  METAD Agricultural Development plc   
7  September 2nd, 2011         Ethiopia Commodity Exchange   

                      certification_address  \
0  309fcf77415a3661ae83e027f7e5f05dad786e44   
1  309fcf77415a3661ae83e027f7e5f05dad786e44   
3  309fcf77415a3661ae83e027f7e5f05dad786e44   
4  309fcf77415a3661ae83e027f7e5f05dad786e44   
7  a176532400aebdc345cf3d870f84ed3ecab6249e   

                      certification_contact unit_of_measurement  \
0  19fef5a731de2db57d16da10287413f5f99bc2dd                   m   
1  19fef5a731de2db57d16da10287413f5f99bc2dd                   m   
3  19fef5a731de2db57d16da10287413f5f99bc2dd                   m   
4  19fef5a731de2db57d16da10287413f5f99bc2dd                   m   
7  61bbaf6a9f341e5782b8e7bd3ebf76aac89fe24b                   m   

  altitude_low_meters altitude_high_meters altitude_mean_meters  \
0              1950.0               2200.0               2075.0   
1              1950.0               2200.0               2075.0   
3              1800.0               2200.0               2000.0   
4              1950.0               2200.0               2075.0   
7              1570.0               1700.0               1635.0   

  bag_weight_num  bag_weight_unit  
0           60.0               kg  
1           60.0               kg  
3           60.0               kg  
4           60.0               kg  
7           60.0               kg  

[5 rows x 45 columns]

From all the six countries only United States (Hawaii) is using lbs as a measurable unit. So, we will add lbs 
to the missing entries from United States (Hawaii). In Nicaragua, we have some mix entries but majariyt is 
kg and in that respect we will not change them to lbs.

In [32]:
# List of indexes you want to update
indexes_to_update = [13, 51, 313, 606, 933]

# Set 'lbs' for those rows in the 'bag_weight_unit' column
coffe_index.loc[indexes_to_update, "bag_weight_unit"] = 'lbs'

In [33]:
coffe_index.loc[indexes_to_update, ['bag_weight_num', 'bag_weight_unit']]

bag_weight_num bag_weight_unit
13              1.0             lbs
51              2.0             lbs
313             1.0             lbs
606             1.0             lbs
933             1.0             lbs

In [34]:
# 4.Convert the lbs to kg

# Ou analysis shows that 90 % of the units are in kg. Taking this into account and according to the data
# in internet for the measurement of coffe bags, we decide to convert all lb and lbs(they both means pounds)
# to kilograms.

coffe_index["bag_weight_kg"] = np.where(
    coffe_index["bag_weight_unit"].isin(['lb', 'lbs']),
    coffe_index["bag_weight_num"]* 0.453592,
    coffe_index["bag_weight_num"]
)

In [35]:
coffe_index["bag_weight_kg"].value_counts(dropna=False) 

bag_weight_kg
1.000000        335
60.000000       256
69.000000       200
70.000000       156
2.000000        123
45.359200        59
30.000000        29
2.267960         23
6.000000         21
20.000000        14
50.000000        14
0.453592         12
10.000000        11
59.000000        10
1.360776          7
5.000000          7
0.907184          6
18975.000000      4
1.814368          4
80.000000         4
0.000000          4
46.000000         3
29.000000         2
19200.000000      2
12000.000000      2
35.000000         2
66.000000         2
40.000000         2
25.000000         2
15.000000         2
9000.000000       2
8.000000          1
3.000000          1
350.000000        1
24.947560         1
67.000000         1
4.000000          1
24.000000         1
100.000000        1
36.287360         1
660.000000        1
1500.000000       1
13800.000000      1
1218.000000       1
34.000000         1
18.000000         1
68.038800         1
18000.000000      1
59.874144         1
58.966

In [36]:
# Delete the other columns

### Problem 5. Dates (1 point)
This should remind you of problem 4 but it's slightly nastier. Fix the harvest years, document the process.

While you're here, fix the expiration dates, and grading dates. Unlike the other column, these should be dates (`pd.to_datetime()` is your friend).

In [38]:
coffe_index.harvest_year.value_counts(dropna=False)

harvest_year
2012                        354
2014                        233
2013                        181
2015                        129
2016                        124
2017                         70
NaN                          47
2013/2014                    29
2015/2016                    28
2011                         26
2017 / 2018                  19
2014/2015                    19
2009/2010                    12
2010                         10
2010-2011                     6
2016 / 2017                   6
4T/10                         4
March 2010                    3
Mayo a Julio                  3
2009-2010                     3
4T/2010                       3
Abril - Julio                 2
January 2011                  2
2011/2012                     2
08/09 crop                    2
TEST                          1
mmm                           1
4T72010                       1
2018                          1
1t/2011                       1
2016/2017                  

In [39]:
coffe_index.harvest_year.dtypes

dtype('O')

#### Problems found with harvest_year column
Similat to problem 4, we look at value_counts(), the dtypes and the column data. The analysis shows that we have the following mix of messy data:  
- single years - 2011, 2012
- years ranges - 2016/2017, 2013/2014   
- years with text periods - January 2011
- periods with only text - Abril - Julio 
- data with quarter indicators - 4t/2010. 1T/2011 
- invalid or unknown data - TEST, mmm 

Our goal is to have usable numerical format consisting of single representative year.
To obtain this we will make the following assumptions:
1. We have to make a desicion which year to take into considerations from the mix year ranges. It is realy hard to decide but taking into account that the best coffee is picked in hte middle
(source: https://coffeehunter.com/coffee-seasonality/; https://driftaway.coffee/when-is-coffee-harvested/) and by the end of the harvest season we will assume that the harvest year is the second year from years ranges.
2. For the data with quarter indicators 4t/2010. 1T/2011 ( in spanish and portugeass T indicates "Trimestre" - Quater) we will leave just the year.
3. For periods only with text - we will replacethem with Unknown      


In [42]:
# Develop a function to clean the data 

def extract_clean_year(value):
    if pd.isna(value):
        return "Unknown"
    value = str(value).lower().strip()

   # Fix malformed: extract last 4 digits if there's a T followed by 4+ digits
    malformed_match = re.match(r'.*t.*(\d{4})$', value)
    if malformed_match:
        return malformed_match.group(1)

    # Standard YYYY
    match = re.match(r'^\d{4}$', value)
    if match:
        return value

    # YYYY/YYYY or YYYY-YYYY or YYYY / YYYY
    match = re.search(r'(\d{4})\s*[/\-]\s*(\d{4})', value)
    if match:
        return match.group(2)

    # YY/YY like 08/09 crop
    match = re.search(r'(\d{2})/(\d{2})', value)
    if match:
        return f"20{match.group(2)}"

    # Fallback: any 4-digit year
    match = re.search(r'(\d{4})', value)
    if match:
        return match.group(1)

    return "Unknown" 

coffe_index["harvest_year_clean"]= coffe_index["harvest_year"].apply(extract_clean_year)

In [43]:
coffe_index["harvest_year_clean"].value_counts(dropna=False)

harvest_year_clean
2012       356
2014       262
2013       181
2016       152
2015       148
2017        77
Unknown     61
2011        40
2010        37
2018        20
2009         4
2020         1
Name: count, dtype: int64

In [44]:
coffe_index.iloc[948]

species                                                   Arabica
owner                              exportadora de cafe condor s.a
country_of_origin                                        Colombia
farm_name                                                     NaN
lot_number                                                    NaN
mill                        trilladora bonanza - armenia quindioa
ico_number                                              3-68-0130
company                                                       NaN
altitude                                                1880 msnm
region                                                      cauca
producer                                                      NaN
number_of_bags                                                250
bag_weight                                                  70 kg
in_country_partner                                       Almacafé
harvest_year                                              4T72010
grading_da

In [45]:
coffe_index.columns

Index(['species', 'owner', 'country_of_origin', 'farm_name', 'lot_number',
       'mill', 'ico_number', 'company', 'altitude', 'region', 'producer',
       'number_of_bags', 'bag_weight', 'in_country_partner', 'harvest_year',
       'grading_date', 'owner_1', 'variety', 'processing_method', 'aroma',
       'flavor', 'aftertaste', 'acidity', 'body', 'balance', 'uniformity',
       'clean_cup', 'sweetness', 'cupper_points', 'total_cup_points',
       'moisture', 'category_one_defects', 'quakers', 'color',
       'category_two_defects', 'expiration', 'certification_body',
       'certification_address', 'certification_contact', 'unit_of_measurement',
       'altitude_low_meters', 'altitude_high_meters', 'altitude_mean_meters',
       'bag_weight_num', 'bag_weight_unit', 'bag_weight_kg',
       'harvest_year_clean'],
      dtype='object')

In [46]:
coffe_index.expiration.value_counts()

expiration
July 11th, 2013        25
December 26th, 2014    25
June 6th, 2013         19
August 30th, 2013      18
July 26th, 2013        15
                       ..
March 8th, 2012         1
May 11th, 2012          1
December 1st, 2012      1
April 27th, 2013        1
December 23rd, 2015     1
Name: count, Length: 566, dtype: int64

In [47]:
coffe_index.expiration.value_counts(dropna=False)

expiration
July 11th, 2013        25
December 26th, 2014    25
June 6th, 2013         19
August 30th, 2013      18
July 26th, 2013        15
                       ..
March 8th, 2012         1
May 11th, 2012          1
December 1st, 2012      1
April 27th, 2013        1
December 23rd, 2015     1
Name: count, Length: 566, dtype: int64

In [135]:
# Remove ordinal suffixes using regex
def clean_date(date_str):
    if pd.isna(date_str):
        return None
    # Strip whitespace and newlines
    date_str = date_str.strip()
    # Remove 'st', 'nd', 'rd', 'th' from day part
    return re.sub(r'(\d{1,2})(st|nd|rd|th)', r'\1', date_str)

# Apply the cleaning function
coffe_index["expiration_clean"] = coffe_index["expiration"].apply(clean_date)

# Convert the cleaned expiration column to datetime
coffe_index["expiration_clean"] = pd.to_datetime(coffe_index["expiration_clean"], errors="coerce")

# Apply the cleaning function
coffe_index["granding_date_clean"] = coffe_index["grading_date"].apply(clean_date)

# Convert the cleaned expiration column to datetime
coffe_index["granding_date_clean"] = pd.to_datetime(coffe_index["granding_date_clean"], errors="coerce")

In [137]:
coffe_index["expiration_clean"].isna().sum()

0

In [139]:
coffe_index.grading_date.value_counts(dropna=False)

grading_date
July 11th, 2012        25
December 26th, 2013    24
June 6th, 2012         19
August 30th, 2012      18
July 26th, 2012        15
                       ..
March 9th, 2011         1
May 12th, 2011          1
December 2nd, 2011      1
April 27th, 2012        1
December 23rd, 2014     1
Name: count, Length: 567, dtype: int64

In [143]:
coffe_index["granding_date_clean"].isna().sum()

0

### Problem 6. Countries (1 point)
How many coffees are there with unknown countries of origin? What can you do about them?

In [51]:
coffe_index.country_of_origin.value_counts()

country_of_origin
Mexico                          236
Colombia                        183
Guatemala                       181
Brazil                          132
Taiwan                           75
United States (Hawaii)           73
Honduras                         53
Costa Rica                       51
Ethiopia                         44
Tanzania, United Republic Of     40
Uganda                           36
Thailand                         32
Nicaragua                        26
Kenya                            25
El Salvador                      21
Indonesia                        20
China                            16
India                            14
Malawi                           11
United States                    10
Peru                             10
Myanmar                           8
Vietnam                           8
Haiti                             6
Philippines                       5
United States (Puerto Rico)       4
Panama                            4
Ecuador   

In [52]:
coffe_index.country_of_origin.value_counts(dropna=False)

country_of_origin
Mexico                          236
Colombia                        183
Guatemala                       181
Brazil                          132
Taiwan                           75
United States (Hawaii)           73
Honduras                         53
Costa Rica                       51
Ethiopia                         44
Tanzania, United Republic Of     40
Uganda                           36
Thailand                         32
Nicaragua                        26
Kenya                            25
El Salvador                      21
Indonesia                        20
China                            16
India                            14
Malawi                           11
United States                    10
Peru                             10
Myanmar                           8
Vietnam                           8
Haiti                             6
Philippines                       5
United States (Puerto Rico)       4
Panama                            4
Ecuador   

In [55]:
coffe_index.country_of_origin.isna().value_counts()

country_of_origin
False    1338
True        1
Name: count, dtype: int64

In [56]:
unknown_countries = coffe_index[coffe_index["country_of_origin"].isna()]

In [57]:
unknown_countries

species               owner country_of_origin farm_name lot_number mill  \
1197  Arabica  racafe & cia s.c.a               NaN       NaN        NaN  NaN   

     ico_number company altitude region  ...  \
1197  3-37-1980     NaN      NaN    NaN  ...   

                         certification_contact  unit_of_measurement  \
1197  70d3c0c26f89e00fdae6fb39ff54f0d2eb1c38ab                    m   

     altitude_low_meters altitude_high_meters altitude_mean_meters  \
1197                 NaN                  NaN                  NaN   

     bag_weight_num bag_weight_unit bag_weight_kg harvest_year_clean  \
1197           70.0              kg          70.0            Unknown   

      expiration_clean  
1197        2012-02-29  

[1 rows x 48 columns]

We know who is the owner of that coffe and we will try to find the country of origin via the owner

In [59]:
coffe_index[coffe_index["owner"] == "racafe & cia s.c.a"]

species               owner country_of_origin farm_name lot_number  \
97    Arabica  racafe & cia s.c.a          Colombia       NaN  3-37-1829   
106   Arabica  racafe & cia s.c.a          Colombia       NaN  3-37-0277   
107   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
110   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
145   Arabica  racafe & cia s.c.a          Colombia       NaN  3-37-1724   
146   Arabica  racafe & cia s.c.a          Colombia       NaN  3-37-1291   
151   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
174   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
183   Arabica  racafe & cia s.c.a          Colombia       NaN  3-37-1704   
197   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
198   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
211   Arabica  racafe & cia s.c.a          Colombia       NaN  3-37-3825   
216   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
235   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
249   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
268   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
275   Arabica  racafe & cia s.c.a          Colombia       NaN  3-37-2112   
299   Arabica  racafe & cia s.c.a          Colombia       NaN  3-37-1510   
304   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
325   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
337   Arabica  racafe & cia s.c.a          Colombia       NaN  3-37-2548   
339   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
347   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
356   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
371   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
397   Arabica  racafe & cia s.c.a          Colombia       NaN  3-37-3729   
398   Arabica  racafe & cia s.c.a          Colombia       NaN  3-37-2623   
436   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
455   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
467   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
479   Arabica  racafe & cia s.c.a          Colombia       NaN  3-37-0278   
492   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
513   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
517   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
525   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
527   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
535   Arabica  racafe & cia s.c.a          Colombia       NaN  3-37-4468   
549   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
557   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
571   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
592   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
603   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
632   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
651   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
671   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
679   Arabica  racafe & cia s.c.a          Colombia       NaN  3-37-4434   
691   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
753   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
755   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
805   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
814   Arabica  racafe & cia s.c.a          Colombia       NaN        NaN   
850   Arabica  racafe & cia s.c.a          Colombia   

The country of the origin of the owner is Columbia. If we look at the ico_number there is also match. 

In [ ]:
# We will replace the NaN with Columbia


### Problem 7. Owners (1 point)
There are two suspicious columns, named `Owner`, and `Owner.1` (they're likely called something different after you solved problem 3). Do something about them. Is there any link to `Producer`?

In [62]:
coffe_index.owner.value_counts()

owner
juan luis alvarado romero           155
racafe & cia s.c.a                   60
exportadora de cafe condor s.a       54
kona pacific farmers cooperative     52
ipanema coffees                      50
                                   ... 
alvaro quiros perez                   1
olivia hernandez virves               1
finca las nieves                      1
pedro santos e silva                  1
james moore                           1
Name: count, Length: 315, dtype: int64

In [63]:
coffe_index.owner.info()

<class 'pandas.core.series.Series'>
Index: 1339 entries, 0 to 1338
Series name: owner
Non-Null Count  Dtype 
--------------  ----- 
1332 non-null   object
dtypes: object(1)
memory usage: 53.2+ KB


In [64]:
coffe_index.owner_1.value_counts()

owner_1
Juan Luis Alvarado Romero           155
Racafe & Cia S.C.A                   60
Exportadora de Cafe Condor S.A       54
Kona Pacific Farmers Cooperative     52
Ipanema Coffees                      50
                                   ... 
ALVARO QUIROS PEREZ                   1
OLIVIA HERNANDEZ VIRVES               1
FINCA LAS NIEVES                      1
Pedro Santos e Silva                  1
James Moore                           1
Name: count, Length: 319, dtype: int64

In [65]:
# Compare the two columns 
# First we will make the text in the two columns lowercase and strip whitespace for comparision
coffe_index["owner_clean"]= coffe_index["owner"].str.lower().str.strip()
coffe_index["owner_1_clean"]= coffe_index["owner_1"].str.lower().str.strip()

In [66]:
coffe_index.owner_clean.value_counts() 

owner_clean
juan luis alvarado romero           155
racafe & cia s.c.a                   60
exportadora de cafe condor s.a       54
kona pacific farmers cooperative     52
ipanema coffees                      50
                                   ... 
alvaro quiros perez                   1
olivia hernandez virves               1
finca las nieves                      1
pedro santos e silva                  1
james moore                           1
Name: count, Length: 315, dtype: int64

In [67]:
coffe_index.owner_1_clean.value_counts()

owner_1_clean
juan luis alvarado romero           155
racafe & cia s.c.a                   60
exportadora de cafe condor s.a       54
kona pacific farmers cooperative     52
ipanema coffees                      50
                                   ... 
romulo bello flores                   1
rachel peterson                       1
josé luis rojas yeo                   1
nitin coffee estate                   1
james moore                           1
Name: count, Length: 317, dtype: int64

In [68]:
# Compare the clean columns
coffe_index["compare"] = coffe_index["owner_clean"] == coffe_index["owner_1_clean"]

In [69]:
# See in how many values differ
coffe_index["compare"].value_counts()

compare
True     1328
False      11
Name: count, dtype: int64

In [70]:
# List the rows which differ
differences = coffe_index[coffe_index["compare"] == False][["owner_clean", "owner_1_clean"]]

In [71]:
differences

owner_clean                     owner_1_clean
219                         ceca, s.a.                         ceca,s.a.
364                                NaN                               NaN
392   federacion nacional de cafeteros  federación nacional de cafeteros
459                                NaN                               NaN
602                                NaN                               NaN
734                      klem organics                      klemorganics
848                                NaN                               NaN
882                                NaN                               NaN
961                      klem organics                      klemorganics
975                                NaN                               NaN
1044                               NaN                               NaN

Taking into account that the anasyss shows differences between the two columns are 
related to small ... we can remove column Owner   

In [73]:
owner_set = set(coffe_index['owner_clean'].dropna().unique())
owner1_set = set(coffe_index['owner_1_clean'].dropna().unique())

only_in_owner = owner_set - owner1_set
only_in_owner1 = owner1_set - owner_set

In [74]:
only_in_owner

{'klem organics'}

In [75]:
only_in_owner1

{'ceca,s.a.', 'federación nacional de cafeteros', 'klemorganics'}

In [149]:
coffe_index.producer.value_counts()

producer
La Plata                           30
Ipanema Agrícola SA                22
Doi Tung Development Project       17
VARIOS                             12
Ipanema Agricola                   12
                                   ..
Kucofa farmers group                1
GRUPO MEDINA                        1
COFFEE COMPANY / CAFECOL            1
ORLANDO ORTOZ / HIPOLITO TORRES     1
Cafe Cazengo                        1
Name: count, Length: 692, dtype: int64

To find if there are any link to Producers column, first we will make the text in the column lowercase and strip the white speces  

In [151]:
coffe_index["producer_clean"]= coffe_index["producer"].str.lower().str.strip()

In [153]:
# Compare the two
coffe_index["relation"] = coffe_index["producer_clean"] == coffe_index["owner_clean"]

In [155]:
coffe_index.relation.value_counts()

relation
False    1188
True      151
Name: count, dtype: int64

In [157]:
coffe_index[coffe_index["relation"]]

species                              owner country_of_origin  \
0     Arabica                          metad plc          Ethiopia   
1     Arabica                          metad plc          Ethiopia   
4     Arabica                          metad plc          Ethiopia   
9     Arabica             diamond enterprise plc          Ethiopia   
22    Arabica              roberto licona franco            Mexico   
...       ...                                ...               ...   
1314  Robusta                             ugacof            Uganda   
1315  Robusta       katuka development trust ltd            Uganda   
1324  Robusta  kasozi coffee farmers association            Uganda   
1328  Robusta                 kawacom uganda ltd            Uganda   
1330  Robusta              mannya coffee project            Uganda   

                               farm_name lot_number                      mill  \
0                              metad plc        NaN                 metad plc   
1                              metad plc        NaN                 metad plc   
4                              metad plc        NaN                 metad plc   
9                      tulla coffee farm        NaN         tulla coffee farm   
22                          la herradura        NaN              la herradura   
...                                  ...        ...                       ...   
1314                 ugacof project area        NaN                    ugacof   
1315  katikamu capca farmers association        NaN  katuka development trust   
1324               kasozi coffee farmers        NaN                       NaN   
1328                            bushenyi        NaN                   kawacom   
1330               mannya coffee project        NaN     mannya coffee project   

     ico_number                            company   altitude  \
0     2014/2015  metad agricultural developmet plc  1950-2200   
1     2014/2015  metad agricultural developmet plc  1950-2200   
4     2014/2015  metad agricultural developmet plc  1950-2200   
9       2014/15             diamond enterprise plc  1795-1850   
22            0                                NaN       1320   
...         ...                                ...        ...   
1314          0                         ugacof ltd       1212   
1315          0       katuka development trust ltd  1200-1300   
1324          0  kasozi coffee farmers association       1367   
1328          0                 kawacom uganda ltd       1600   
1330          0              mannya coffee project       1200   

                           region  ... bag_weight_unit  bag_weight_kg  \
0                    guji-hambela  ...              kg           60.0   
1                    guji-hambela  ...              kg           60.0   
4                    guji-hambela  ...              kg           60.0   
9     snnp/kaffa zone,gimbowereda  ...              kg           60.0   
22                         xalapa  ...              kg            1.0   
...                           ...  ...             ...            ...   
1314                      central  ...              kg           60.0   
1315        luwero central region  ...              kg           60.0   
1324                      eastern  ...              kg           60.0   
1328                      western  ...              kg           60.0   
1330                     southern  ...              kg           60.0   

     harvest_year_clean expiration_clean                        owner_clean  \
0                  2014       2016-04-03                          metad plc   
1                  2014       2016-04-03                          metad plc   
4                  2014       2016-04-03                          metad plc   
9                  2014       2016-03-29             diamond enterprise plc   
22                 2012       2013-07-26              roberto licona franco   
...                 ...              ...                    

The data show that in 151 case the producers and the owners are the same. So in those cases we have a clear relation.

### Problem 8. Coffee color by country and continent (1 point)
Create a table which shows how many coffees of each color are there in every country. Leave the missing values as they are.

**Note:** If you ask me, countries should be in rows, I prefer long tables much better than wide ones.

Now do the same for continents. You know what continent each country is located in.

In [159]:
coffe_index.color.value_counts()

color
Green           870
Bluish-Green    114
Blue-Green       85
Name: count, dtype: int64

In [165]:
coffe_index_color = coffe_index.pivot_table(index = "country_of_origin", columns = "color", values = "species", 
                                            aggfunc = "count")

In [167]:
coffe_index_color.reset_index()

color             country_of_origin  Blue-Green  Bluish-Green  Green
0                            Brazil        14.0          12.0   92.0
1                           Burundi         NaN           NaN    1.0
2                             China         NaN           NaN   16.0
3                          Colombia         8.0           8.0  118.0
4                        Costa Rica        10.0           9.0   28.0
5                     Cote d?Ivoire         NaN           1.0    NaN
6                           Ecuador         2.0           1.0    NaN
7                       El Salvador         2.0           2.0    9.0
8                          Ethiopia         NaN           2.0   15.0
9                         Guatemala         2.0           7.0  159.0
10                            Haiti         1.0           1.0    2.0
11                         Honduras         3.0           2.0   47.0
12                            India         1.0           2.0   10.0
13                        Indonesia         4.0           3.0    6.0
14                            Japan         NaN           NaN    1.0
15                            Kenya         3.0           6.0   14.0
16                             Laos         1.0           1.0    1.0
17                           Malawi         1.0           4.0    6.0
18                        Mauritius         NaN           NaN    1.0
19                           Mexico         7.0           7.0  156.0
20                          Myanmar         NaN           NaN    7.0
21                        Nicaragua         NaN           2.0   19.0
22                           Panama         NaN           1.0    2.0
23                 Papua New Guinea         NaN           1.0    NaN
24                             Peru         NaN           2.0    6.0
25                      Philippines         1.0           1.0    2.0
26                           Rwanda         NaN           1.0    NaN
27                           Taiwan        14.0           5.0   50.0
28     Tanzania, United Republic Of         NaN           7.0   32.0
29                         Thailand         3.0           7.0   10.0
30                           Uganda         2.0           1.0   32.0
31                    United States         NaN           NaN    1.0
32           United States (Hawaii)         6.0          16.0   21.0
33      United States (Puerto Rico)         NaN           1.0    3.0
34                          Vietnam         NaN           1.0    3.0

In [169]:
coffe_index_color.columns

Index(['Blue-Green', 'Bluish-Green', 'Green'], dtype='object', name='color')

In [171]:
coffe_index.columns

Index(['species', 'owner', 'country_of_origin', 'farm_name', 'lot_number',
       'mill', 'ico_number', 'company', 'altitude', 'region', 'producer',
       'number_of_bags', 'bag_weight', 'in_country_partner', 'harvest_year',
       'grading_date', 'owner_1', 'variety', 'processing_method', 'aroma',
       'flavor', 'aftertaste', 'acidity', 'body', 'balance', 'uniformity',
       'clean_cup', 'sweetness', 'cupper_points', 'total_cup_points',
       'moisture', 'category_one_defects', 'quakers', 'color',
       'category_two_defects', 'expiration', 'certification_body',
       'certification_address', 'certification_contact', 'unit_of_measurement',
       'altitude_low_meters', 'altitude_high_meters', 'altitude_mean_meters',
       'bag_weight_num', 'bag_weight_unit', 'bag_weight_kg',
       'harvest_year_clean', 'expiration_clean', 'owner_clean',
       'owner_1_clean', 'compare', 'granding_date_clean', 'producer_clean',
       'relation'],
      dtype='object')

In [179]:
coffe_index.country_of_origin.value_counts()

country_of_origin
Mexico                          236
Colombia                        183
Guatemala                       181
Brazil                          132
Taiwan                           75
United States (Hawaii)           73
Honduras                         53
Costa Rica                       51
Ethiopia                         44
Tanzania, United Republic Of     40
Uganda                           36
Thailand                         32
Nicaragua                        26
Kenya                            25
El Salvador                      21
Indonesia                        20
China                            16
India                            14
Malawi                           11
United States                    10
Peru                             10
Myanmar                           8
Vietnam                           8
Haiti                             6
Philippines                       5
United States (Puerto Rico)       4
Panama                            4
Ecuador   

### Problem 9. Ratings (1 point)
The columns `Aroma`, `Flavor`, etc., up to `Moisture` represent subjective ratings. Explore them. Show the means and range; draw histograms and / or boxplots as needed. You can even try correlations if you want. What's up with all those ratings?

In [ ]:
coffe_index.aroma.value_counts()

In [ ]:
coffe_index.flavor.value_counts()

In [ ]:
coffe_index.moisture.value_counts()

### Problem 10. High-level errors (1 point)
Check the countries against region names, altitudes, and companies. Are there any discrepancies (e.g. human errors, like a region not matching the country)? Take a look at the (cleaned) altitudes; there has been a lot of preprocessing done to them. Was it done correctly?

In [ ]:
coffe_index.altitude.value_counts()

In [ ]:
coffe_index.region.value_counts()

In [ ]:
coffe_index.company

### * Problem 11. Clean and explore at will
The dataset claimed to be clean, but we were able to discover a lot of things to fix and do better.

Play around with the data as much as you wish, and if you find variables to tidy up and clean - by all means, do that!

In [ ]:
pd.set_option('display.max_columns', None)